drivers: pwm: stm32: Add trigger output configuration

embRoobin opened the pull request:
## Summary

This push request enables trigger output (TRGO) configuration for ST pwm instances via devictree.
Therefore the usage of the STM32 interconnect matrix is possible for pwm instances.
API remains unchanged.

## What is changed
- Add new dts-binding to configure source of TRGO signal.
- Add predefined constants for configuration
- Add TRGO configuation during instance init.

## Alternatives

As described in https://github.com/zephyrproject-rtos/zephyr/issues/70097 an interconnect subsystem would solve the problem as well.

In comparison to a new subsystem this solution is more lightweight on code size and most init time. 
Drawback is the hardware dependency to ST devices (but you have to know the hardware interconnect capabilities of your SoC in both cases to archive maximal performance).

erwango commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -699,6 +700,12 @@ static int pwm_stm32_init(const struct device *dev)
+#ifdef IS_TIM_MASTER_INSTANCE
+	if (IS_TIM_MASTER_INSTANCE(timer)) {
+		LL_TIM_SetTriggerOutput(timer, cfg->mastermode);
+	}
```
Might be good to return an error if instance can't operate as master timer ?

embRoobin replied:
You mean something like this?

```c
	if (IS_TIM_MASTER_INSTANCE(timer)) {
		LL_TIM_SetTriggerOutput(timer, cfg->mastermode);
	}else{
		if (cfg->mastermode != 0){
			LOG_ERR("Timer does not support mastermode, but non default value is specified");
			return -EINVAL;
		}
	}
```
Is this the correct error value?

erwango replied:
Yes, indeed. `ENOTSUP` might fit better

erwango commented on dts/bindings/timer/st,stm32-timers.yaml:
```diff
@@ -46,3 +46,32 @@ properties:
 
       If absent, then STM32_TIM_COUNTERMODE_UP is used (reset state).
+
+  st,master-mode:
```
I'm a bit annoyed by the discrepancy vs `st,countermode` writing, even if the latter feels less nice.

erwango replied:
Note also that it should be added to `st,stm32-lptim.yaml` blocklist

erwango reviewed (changes requested):
We miss a proper implementation of this:
- activation on a target
- example of sample using it (could be added in sample/boards/st)

This is probably based on https://github.com/zephyrproject-rtos/zephyr/pull/43977, but IMO we should have pushed the review further on this PR (learning every day).

What should also be clarified:
incompatibility of the provided values on specific series: How to identify them, how to derogate. What it the current base of know compatible devices.

embRoobin commented on dts/bindings/timer/st,stm32-timers.yaml:
```diff
@@ -46,3 +46,32 @@ properties:
+      - 0x40 # STM32_TIM_MASTERMODE_COMPARE_OC1REFC
+      - 0x50 # STM32_TIM_MASTERMODE_COMPARE_OC2REFC
+      - 0x60 # STM32_TIM_MASTERMODE_COMPARE_OC3REFC
+      - 0x70 # STM32_TIM_MASTERMODE_COMPARE_OC4REFC
```
Nice hint, thx :)
I like that it removes the hardcoded dependency into the LL-Lib.
Will be implemented shortly.

etienne-lms reviewed (commented):
Could you squash commit "drivers: pwm: stm32: Refactor to use string enum"
into commit "drivers: pwm: stm32: Add trigger output configuration"?

etienne-lms commented:
Actually, commit "dts: bindings: stm32: Add unsupported entry" should also be squashed.

etienne-lms reviewed (commented):
See also the few CI "Run compliance checks" reports.

etienne-lms reviewed (dismissed):
Non blocking comment. LGTM.

embRoobin committed:
drivers: pwm: stm32: Add trigger output configuration

- Add new dts binding for configurable master mode
- Add exception for LP timers (not supported)
- Add required defines for configuration options
- Implement configuration of master mode during device init

Signed-off-by: Matthias Plöger <matthias.ploeger@phoenixcontact.com>

embRoobin committed:
samples: boards: stm32: pwm mastermode

Add an sample for usage of the timer/pwm mastermode configuration.
Timer is configured to generate 2 TRGO events per pwm period to the
interconnect matrix.

Signed-off-by: Matthias Plöger <matthias.ploeger@phoenixcontact.com>

djiatsaf-st reviewed (dismissed):
+1  for CI testing

mathieuchopstm reviewed (commented):
otherwise LGTM

etienne-lms commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -700,6 +701,17 @@ static int pwm_stm32_init(const struct device *dev)
+			return -ENOTSUP;
+		}
+	}
+#endif /* IS_TIM_MASTER_INSTANCE*/
```
Also reject non 0 values when `IS_TIM_MASTER_INSTANCE` is not defined, as done with `IS_TIM_BREAK_INSTANCE`/`st-deadtime`?
```suggestion
#else
	if (cfg->mastermode != 0) {
		LOG_ERR("%s: Timer does not support mastermode", dev->name);
		return -ENOTSUP;
	}
#endif /* IS_TIM_MASTER_INSTANCE*/

etienne-lms commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -785,6 +797,7 @@ static void pwm_stm32_irq_config_func_##index(const struct device *dev)		\
 		.prescaler = DT_PROP(PWM(index), st_prescaler),		       \
 		.countermode = DT_PROP(PWM(index), st_countermode),	       \
 		.deadtime = DT_PROP(PWM(index), st_deadtime),		       \
+		.mastermode = CONCAT(LL_TIM_TRGO_, DT_STRING_TOKEN(PWM(index), st_mastermode)),	\
```
I see there is 1 SoC series for which `LL_TIM_TRGO_RESET` is not defined and that supports `"st,stm32-pwm"` compatible driver: STM32WB0.
```suggestion
#ifdef LL_TIM_TRGO_RESET
#define STM32_LL_HAS_TIM_TRGO_RESET 1
#else
#define STM32_LL_HAS_TIM_TRGO_RESET 0
#endif

(...)

		.mastermode = COND_CODE_1(STM32_LL_HAS_TIM_TRGO_RESET,	       \
					  CONCAT(LL_TIM_TRGO_,		       \
						 DT_STRING_TOKEN(PWM(index),   \
						 st_mastermode)),	       \
					  (0)),				       \
```
or like.

mathieuchopstm reviewed (commented):
LGTM besides Etienne's remarks and:

etienne-lms commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -33,9 +33,12 @@ LOG_MODULE_REGISTER(pwm_stm32, CONFIG_PWM_LOG_LEVEL);
+#define HAS_MASTERMODE_SUPPORT 1
+#else
+#define HAS_MASTERMODE_SUPPORT 0
+#define IS_TIM_MASTER_INSTANCE(INSTANCE) 0
```
IMHO, this change (and the related one around line 806) should be squashed into commit
*"drivers: pwm: stm32: Add trigger output configuration"*.

etienne-lms commented on samples/boards/st/timer_mastermode/sample.yaml:
```diff
@@ -0,0 +1,17 @@
+      - pwm
+    filter: dt_compat_enabled("st,stm32-timers")
+    platform_allow:
+      - nucleo_g070rb
```
Note that **boards/st/nucleo_g070rb/nucleo_g070rb.yaml** does not set supported `pm` tag hence this test will not be triggered.

erwango reviewed (changes requested):
Thanks for the changes!

One last request on my side:
Please align on other ST samples:
From `samples/boards/st/timer/mastermode`
to `samples/boards/st/pwm/mastermode/`
(also use "pwm" instead of "timer", as timers could also be used by counter api, which doesn't implement mastermode today)

mathieuchopstm reviewed (commented):
Otherwise LGTM

embRoobin commented on samples/boards/st/pwm/mastermode/src/main.c:
```diff
@@ -0,0 +1,43 @@
+
+	if (!device_is_ready(pwm)) {
+		LOG_ERR("%s: pwm device not ready", pwm->name);
+		return -ENODEV;
```
> main function is expected to always return 0 because the return value is not tested.
> The LOG_ERR() should be enough.

When looking at other samples ([blinky](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/basic/blinky/src/main.c)) also (default) return values are used. 
The additional error code only returns a non default value.

In addition testing wouldn't fail, if no `return` is present.

>Maybe replace all LOG_ERR() and LOG_INF() with plain printf().

Also other (not many) driver samples are using the logging module.
The benefit when testing is a clear error/information marking on the debug messages.

The question I am asking myself is: Why should we use something with less functionality?

andyross reviewed (approved):
I may not be the best assignee for ST driver stuff, but this has been well-reviewed and looks uncontroversial.
